thunder/torch/__init__.py

```diff
@@ -3489,6 +3489,21 @@ def embedding_backward(grad, indices, num_weights, padding_idx, scale_grad_by_fr
     return result
 
 
+@torchsymbol(torch.nn.functional.one_hot, id="torch.nn.functional.one_hot", is_method=False)
+def one_hot(a: TensorProxy, /, num_classes: int) -> TensorProxy:  # Should we use TensorProxy for a?
+    if num_classes == -1:
+        # TODO: implement when we're ready to support auto inference using `.item()`
+        utils.check(False, lambda: "num_classes=-1 not supported yet.", exception_type=NotImplementedError)
+    # TODO: should we implement this check? I suspect this might hinder performance, also we do not support .any)
+    #  utils.check(a.any() >= 0, lambda f"input tensor should have non-negative values", exception_type=ValueError)
+
+    canvas = zeros(*a.shape, num_classes, dtype=dtypes.int64)  # this is equivalent to the use of .long in torch
+    index = a.unsqueeze(-1)
+    src = ones_like(index, dtype=dtypes.int64)
+
+    return scatter_add(canvas, dim=-1, index=index, src=src)
```

Comment:
Seems fine for now, but what happens is that we create a tensor full of ones. We could:
- create a tensor with a single 1 and broadcast it.
- leave it be in hopes that sometime in the future our backends/interpreters will improve this bit (when scatter_add will be there for nvFuser, for example). Worth adding a comment, I guess?

Reply:
Hey @nikitaved! Not sure I understand the implementation suggested in _..and broadcast it_. Tried to implement naively by replacing,
'src = ones_like(index, device=a.device, dtype=dtypes.int64)'
with,
`src = tensor([1], device=a.device, dtype=dtypes.int64)` 
and got,
```
thunder/core/prims.py:3009: in scatter_add_meta
    utils.check(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cond = False, s = <function scatter_add_meta.<locals>.<lambda> at 0x7bb2a8c043a0>
exception_type = <class 'RuntimeError'>

    def check(cond: bool, s: Callable[[], str], exception_type: type[Exception] = RuntimeError) -> None:
        """Helper function for raising an error_type (default: RuntimeError) if a boolean condition fails.
    
        s is a callable producing a string to avoid string construction if the error check is passed.
        """
        if not cond:
>           raise exception_type(s())
E           RuntimeError: Expected index (rank=3) to have the same rank as value (rank=1)

thunder/core/baseutils.py:103: RuntimeError
```

Reply:
You can read about broadcasting here: https://numpy.org/doc/stable/user/basics.broadcasting.html

Reply:
Cool, and should this be more efficient operation than current implementation?

Reply:
Well, yes. It will spare us launching a kernel that fills memory with `1`s. This might become redundant once the target executor can fuse together `ones` and `scatter_add`. In the context of NVFuser, this means putting the code of `ones` and `scatter_add` into a single CUDA kernel. Do not worry for now if you do not understand these things, this could be done in some follow-up. You can learn more about executors from the documentation that should be, alas, built locally.